### DISCUSSION TITLE
Can some one help me with implementing the forgot password feature?

### DISCUSSION THREAD

**Discussion by martijnvandenboom** [remix-run/remix#5101] [Q&A]
A warm hello to everyone in this community! Currently I am working on user registration with an e-mail address. The basics are working but the only thing missing now is the forgot password link and mechanism that will send an e-mail to the user with a link to a page where it will allow the user to actually enter a new password and save it to his or her account. Does someone have a tutorial for that? Or is anyone interested to help me out? My project is in Github maybe we can collaborate?

**Comment by machour** [comment]
Hi @martijnvandenboom 👋

Usually, you would implement this feature like this:
- add a `password_reset_token` to your user table. It's a nullable string field. 
- when a user ask for a password reset, ask for the email
- if the email match one of the records of your database, generate a random token and store it in your new field for that user
- send an email to the user with a link to reset password, with that token on the query string (ie `auth/reset-password?token=${generated token}`)
- on that route loader, read the token and verify that it exists in your user table. If it doesn't not, throw an error. Return that token to your component. 
- have a form with the new password field, and a hidden field with the token. 
- when the form is submitted, on the route action, if the token still match a valid user, update the password and set the token to null. 

That's the gist of it, let me know if you have more questions 🙌
